src/analyses/call_graph.cpp

```diff
@@ -51,50 +46,111 @@ void call_grapht::add(
+  {
+    node_indext node_index = add_node();
+    nodes[node_index].function_name = callee;
+  }
+
+  add_edge(caller_idx, callee_idx);
 }
 
-/// Returns an inverted copy of this call graph
-/// \return Inverted (callee -> caller) call graph
-call_grapht call_grapht::get_inverted() const
+
+void call_grapht::output_dot_node(std::ostream &out, node_indext n) const
 {
-  call_grapht result;
-  for(const auto &caller_callee : graph)
-    result.add(caller_callee.second, caller_callee.first);
-  return result;
+  const nodet &node = nodes.at(n);
+
+  for(const auto &edge : node.out)
+  {
+    out << "  \"" << node.function_name << "\" -> " << "\""
+        << nodes[edge.first].function_name << "\" " << " [arrowhead=\"vee\"];"
+        << "\n";
+  }
 }
 
-void call_grapht::output_dot(std::ostream &out) const
+void call_grapht::output_xml_node(std::ostream &out, node_indext n) const
 {
-  out << "digraph call_graph {\n";
+  const nodet &node = nodes.at(n);
 
-  for(const auto &edge : graph)
+  for(const auto &edge : node.out)
   {
-    out << "  \"" << edge.first << "\" -> "
-        << "\"" << edge.second << "\" "
-        << " [arrowhead=\"vee\"];"
-        << "\n";
+    out << "<call_graph_edge caller=\"";
+    xmlt::escape_attribute(id2string(node.function_name), out);
+    out << "\" callee=\"";
+    xmlt::escape_attribute(id2string(nodes[edge.first].function_name), out);
+    out << "\">\n";
   }
+}
 
-  out << "}\n";
+void call_grapht::output_xml(std::ostream &out) const
+{
+  for(node_indext n = 0; n < nodes.size(); n++)
+    output_xml_node(out, n);
 }
 
-void call_grapht::output(std::ostream &out) const
+call_grapht call_grapht::get_inverted() const
 {
-  for(const auto &edge : graph)
+  call_grapht result;
+  for(const auto &n : nodes)
   {
-    out << edge.first << " -> " << edge.second << "\n";
+    for(const auto &i : n.in)
+      result.add(n.function_name, nodes[i.first].function_name);
   }
+  return result;
 }
 
-void call_grapht::output_xml(std::ostream &out) const
+std::list<irep_idt>call_grapht::shortest_function_path
+(irep_idt src, irep_idt dest)
 {
-  for(const auto &edge : graph)
+  std::list<irep_idt> result;
+  node_indext src_idx, dest_idx;
+  if(!get_node_index(src, src_idx))
+    throw "unable to find src function in call graph";
+  if(!get_node_index(dest, dest_idx))
+    throw "unable to find destination function in call graph";
+
+  patht path;
+  shortest_path(src_idx, dest_idx, path);
+  for(const auto &n : path)
   {
-    out << "<call_graph_edge caller=\"";
-    xmlt::escape_attribute(id2string(edge.first), out);
-    out << "\" callee=\"";
-    xmlt::escape_attribute(id2string(edge.second), out);
-    out << "\">\n";
+    result.push_back(nodes[n].function_name);
+  }
+  return result;
+}
+
+
+std::unordered_set<irep_idt, irep_id_hash>
+call_grapht::reachable_functions(irep_idt start_function)
+{
+  std::unordered_set<irep_idt, irep_id_hash> result;
+  std::list<node_indext> worklist;
```

Comment:
Will reimplement the search. 

If I put get_reachable into grapht, it will then return a list of node indexes, and I will need to add a function in call_grapht that interfaces to that function, converting the node indexes to function names. Still ok?

Reply:
Yes that sounds right